### PR TITLE
PCX load screen of campaign

### PR DESCRIPTION
**Pull request by ststl-s**
In `missionmd.ini`, `LS800BkgdName` also can set as a pcx file.

### CODE REVIEW

File: src/Misc/Hooks.PCX.cpp [hunk 1]
```diff
@@ -40,6 +41,44 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
 	return 0;
 }
 
+DEFINE_HOOK(0x552F81, PCX_LoadingScreen_Campaign, 0x5)
+{
+	GET(LoadProgressManager*, pThis, EBP);
+
+	DSurface* pSurface = static_cast<DSurface*>(pThis->ProgressSurface);
+	std::string fileName = ScenarioClass::Instance->LS800BkgdName;
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:49]: No need to construct a std::string object here.

File: src/Misc/Hooks.PCX.cpp [hunk 2]
```diff
@@ -40,6 +41,44 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
...
+
+	//Debug::Log("[LS800Bkgd] file[%s]\n", fileName.c_str());
+
+	if (fileName.length() < 4U)
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:53]: `strlen`
Also, it shouldn't arise a fatal error here. It's better to just return 0 here.

File: src/Misc/Hooks.PCX.cpp [hunk 3]
```diff
@@ -40,6 +41,44 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
...
+		Debug::FatalErrorAndExit("[LS800BkgdName] illegal filename [%s]\n", fileName.c_str());
+
+	if (_strcmpi(fileName.substr(fileName.length() - 4).c_str(), ".pcx") == 0)
+	{
+		PCX::Instance->LoadFile(fileName.c_str());
+		BSurface* pcx = PCX::Instance->GetSurface(fileName.c_str());
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:59]: `if(auto pPCX = PCX::Instance->GetSurface(lpFileName))`

File: src/Misc/Hooks.PCX.cpp [hunk 4]
```diff
@@ -40,6 +41,44 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
...
+		{
+			RectangleStruct surfBounds = { 0, 0, pSurface->Width, pSurface->Height };
+			RectangleStruct pcxBounds = { 0, 0, pcx->Width, pcx->Height };
+
+			RectangleStruct destClip = { 0, 0, pcx->Width, pcx->Height };
+			destClip.X = (pSurface->Width - pcx->Width) / 2;
+			destClip.Y = (pSurface->Height - pcx->Height) / 2;
+
+			PCX::Instance->BlitToSurface(&destClip, pSurface, pcx);
+			//pSurface->CopyFrom(&surfBounds, &destClip, pcx, &pcxBounds, &pcxBounds, true, true);
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:71]: Delete unused code here

File: src/Misc/Hooks.PCX.cpp [hunk 5]
```diff
@@ -40,6 +41,44 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
...
+		Debug::FatalErrorAndExit("[LS800BkgdName] illegal filename [%s]\n", fileName.c_str());
+
+	if (_strcmpi(fileName.substr(fileName.length() - 4).c_str(), ".pcx") == 0)
+	{
+		PCX::Instance->LoadFile(fileName.c_str());
+		BSurface* pcx = PCX::Instance->GetSurface(fileName.c_str());
+
+		if (pcx)
+		{
+			RectangleStruct surfBounds = { 0, 0, pSurface->Width, pSurface->Height };
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:63]: Remove this unused var here.

**Review by secsome — CHANGES_REQUESTED** [review]
Fix it and do some test on it.

**Comment by ststl-s** [comment]
image

**Comment by ststl-s** [comment]
```ini
[SOV01UMD.MAP]
Briefing=Brief:Sov01md
UIName=Name:Sov01md
LSLoadMessage=LoadMsg:Sov01md
LSLoadBriefing=LoadBrief:Sov01md
LS640BriefLocX=20
LS640BriefLocY=20
LS800BriefLocX=20
LS800BriefLocY=20
LS640BkgdName=test2.pcx;LS640S01.SHP
LS800BkgdName=test2.pcx;LS800S01.SHP
LS800BkgdPal=LS800S01.PAL
```

**Comment by ststl-s** [comment]
[test2.zip](https://github.com/Phobos-developers/Phobos/files/9077705/test2.zip)

**Comment by Belonit** [comment]
Question. This feature work with cncnet5.dll if the game resolution is greater than 800x600 ?

**Comment by ststl-s** [comment]
> Question: does this feature work with `cncnet5.dll` if the game resolution is greater than 800x600?

1600x900
$V3}N_VZ`VEYTF210BHPU H

PCX size larger than resolution will cause FatalError, `pSurface->CopyFrom` will not....

File: src/Misc/Hooks.PCX.cpp [hunk 6]
```diff
@@ -40,6 +41,36 @@ DEFINE_HOOK(0x5535D0, PCX_LoadScreen, 0x6)
...
+	DSurface* pSurface = static_cast<DSurface*>(pThis->ProgressSurface);
+	char filename[0x40];
+	strcpy_s(filename, ScenarioClass::Instance->LS800BkgdName);
+	_strlwr_s(filename);
+
+	if (strstr(filename, ".pcx"))
+	{
+		PCX::Instance->LoadFile(filename);
+
+		if (auto pPCX = PCX::Instance->GetSurface(filename))
```
Comment (secsome) [inline src/Misc/Hooks.PCX.cpp:57]: `auto const` here and others looks fine now

**Review by secsome — CHANGES_REQUESTED** [review]

**Comment by secsome** [comment]
Do more tests and this PR can be merged.

**Comment by Metadorius** [comment]
> > Question: does this feature work with `cncnet5.dll` if the game resolution is greater than 800x600?
> 
> 1600x900
> $V3}N_VZ`VEYTF210BHPU H
> 
> PCX size larger than resolution will cause FatalError, `pSurface->CopyFrom` will not....

So does your patch crash if the size is bigger than resolution? Ideal behavior would be allowing it to be centered and cropped.

**Comment by ststl-s** [comment]
> > > Question: does this feature work with `cncnet5.dll` if the game resolution is greater than 800x600?
> > 
> > 
> > 1600x900
> > $V3}N_VZ`VEYTF210BHPU H
> > PCX size larger than resolution will cause FatalError, `pSurface->CopyFrom` will not....
> 
> So does your patch crash if the size is bigger than resolution? Ideal behavior would be allowing it to be centered and cropped.

Now it will not. PCX will show part of it when it is larger than the resolution, and it is centered.

**Comment by ststl-s** [comment]
> > > Question: does this feature work with `cncnet5.dll` if the game resolution is greater than 800x600?
> > 
> > 
> > 1600x900
> > $V3}N_VZ`VEYTF210BHPU H
> > PCX size larger than resolution will cause FatalError, `pSurface->CopyFrom` will not....
> 
> So does your patch crash if the size is bigger than resolution? Ideal behavior would be allowing it to be centered and cropped.

Now it will not. PCX will show part of it when it is larger than the resolution, and it is centered.

**Review by Metadorius — APPROVED** [review]

**Pull request merged**